src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -703,6 +703,8 @@ void ZooKeeper::receiveThread()
             {
                 if (requests_queue.isFinished())
                     break;
+                if (finalization_started.test())
+                    break;
```

Comment:
Why? `CANNOT_READ_ALL_DATA` means that we failed to read Keeper's response from the socket. It does not seem to be related to the finalization

Reply:
finalize firstly calls `shutdown()` on the socket, this will trigger `poll(POLLIN)` (EOF) and the `receiveThread` will trigger a `receiveEvent` which will fail because `read` will return `0` (instead of `4` - length).

Note, that rearrange the code in the finalize (i.e. call `shutdown()` after) since the client sends CLOSE_XID before, and the server may close the connection already, and this will also trigger `poll(POLLIN)` (EOF), and `read` will return `0` also.

Reply:
Got it, thanks. So do we still need the `if (requests_queue.isFinished())` condition?

Reply:
Oh, you are right, could be removed. Rebased.